### ISSUE TITLE
Reform in notebooks w //-comment stripping and param:year format? 

### ISSUE THREAD

**Issue by MattHJensen** [open-source-economics/Tax-Calculator#954]
Many users interact with TC  through a Jupyter Notebook, such as our [10_min_to_tax-calculator](https://github.com/open-source-economics/Tax-Calculator/blob/master/docs/notebooks/10_Minutes_to_Tax-Calculator.ipynb) starting guide. These users are exposed to the [year:param](https://github.com/open-source-economics/Tax-Calculator/blob/master/taxcalc/tests/test_policy.py#L130) style for implementing reforms through `implement_reform()`. 

We are currently adding a capability for TB users to specify a reform in the [param:year style](https://github.com/open-source-economics/Tax-Calculator/blob/master/taxcalc/tests/test_incometaxio.py#L145-L180) available through `read_json_reform_file()`. 

Notebook users _could_ specify a reform in a separate reform file and then read it with `read_json_reform_file()`, but this is unnatural. Most notebook users want everything that they are doing to be displayed in one place so that it is easy to follow their thought process -- that is the purpose of the notebook. 

As things are now, TB users will be writing param:year. Notebook users are writing year:param. 

As one notebook user, I really like the param:year style and I also like the //-comment stripping capabilities that are available through `read_json_reform_file()`; unfortunately, these capabilities are only available if I am ready a file so they are difficult to use from a notebook. 

I am wondering if it is possible to unbundle these capabilities. Maybe the //-comment stripping and param:year capabilities could be added to `implement_reform()`, maybe `read_json_reform_file()` would also accept a string instead of only a filename, or maybe something else. I don't think we want to break the existing capabilities of `implement_reform()`

There could also be a way around this now that I am missing. 

@martinholmer @talumbau

**Comment by martinholmer** [comment]
@MattHJensen said he thought it would be much better for Jupyter Notebook users of the taxcalc package if "JSON reform files" (as read currently read by the `Policy.read_json_reform_file(filename)` static method) could be read from a string specified in a Notebook.

This makes a lot of sense to me; the only question is how best to provide this capability.

In many other taxcalc methods a parameter can have several types and a  different kind of action is taken by the method for each parameter type.

But in this case the current filename parameter is a string and, in the Notebook situation, the JSON reform would be a JSON string (possibly containing //-comments).  So, they would be the same type.

Does it make sense to rename the method as `Policy.read_json_reform(json_input)` (or something like that) where the `json_input` is always a string.  Then if `json_input` named a file that exists, it would be interpreted as a filename and the JSON reform would be read from that file.  If the `json_input` string parameter was not a file that exists, then it would be interpreted as a JSON reform string (possibly including //-comments).

Does this approach make sense?  Is there a better approach?

@MattHJensen @talumbau @zrisher @Amy-Xu @GoFroggyRun @andersonfrailey

**Comment by talumbau** [comment]
I think [explicit is better than implicit](https://www.python.org/dev/peps/pep-0020/). I would suggest: `Policy.read_json_reform(json_input)` always takes a string-like object. Then, I would add:

```

def read_json_reform_from_file(path):
    if os.path.exists(path):
        return Policy.read_json_reform(open(path, 'r').read())
    else:
        raise ValueError("{} does not exist".format(path))
```

**Comment by martinholmer** [comment]
@talumbau said he favored creating **two** different methods to read JSON reforms: one from a file and one from a string.

I'm OK with that, but it goes against the prior approach that I described above.  Just curious why you suggest a different approach in this case when you were the author of the earlier code that takes different action depending on the type of parameter passed to *one* method.

@MattHJensen

**Comment by martinholmer** [comment]
@MattHJensen,  Does merged pull request #958 meet the needs of Notebook users?

**Comment by MattHJensen** [comment]
> @MattHJensen, Does merged pull request #958 meet the needs of Notebook users?

Yes. Thank you very much @martinholmer. I will update the 10-min notebook to take advantage of this feature.
